dali/image/generic_image.cc

```diff
@@ -22,18 +22,24 @@ GenericImage::GenericImage(const uint8_t *encoded_buffer, size_t length, DALIIma
         Image(encoded_buffer, length, image_type) {
 }
 
-
 std::pair<std::shared_ptr<uint8_t>, Image::Shape>
 GenericImage::DecodeImpl(DALIImageType image_type,
                          const uint8_t *encoded_buffer,
                          size_t length) const {
+  const auto shape = PeekShapeImpl(encoded_buffer, length);
+  if (image_type == DALI_ANY_DATA)
+    image_type = shape[2] == 1 ? DALI_GRAY : DALI_RGB;
+  const auto C = IsColor(image_type) ? 3 : 1;
```

Comment:
```suggestion
  if (image_type == DALI_ANY_DATA) {
    const auto shape = PeekShapeImpl(encoded_buffer, length);
    image_type = shape[2] == 1 ? DALI_GRAY : DALI_RGB;
    }
  const auto C = IsColor(image_type) ? 3 : 1;
```
Peek shape is not free.